### ISSUE TITLE
x0vncserver Support CentOS 8

### ISSUE THREAD

**Issue by mddflorida** [TigerVNC/tigervnc#1221]
**Describe the bug**
First of all, I am not convinced this is a bug, but on the other hand with I describe below used to work....
I successfully use X0vncserver under CentOS 7 to make the console remotely accessible via VNC to login and access the desktop using the command: 'x0vncserver -display :0 -PasswordFile /etc/vncpasswd -AlwaysShared -SecurityTypes VncAuth,TLSVnc'

Under CentOS 8 with WaylandEnable=false added to /etc/gdm/custom.conf,  I get:
No protocol specified
x0vncserver: unable to open display ":0"

However, if I configure /etc/gdm/custom.conf with
AutomaticLoginEnable=True
AutomaticLogin=root

to automatically log in root on boot, I get:
Fri Mar 19 09:45:47 2021
 Geometry:    Desktop geometry is set to 1920x1080+0+0
 XDesktop:    Using evdev codemap
 XDesktop:    
 XDesktop:    XTest extension present - version 2.2
 XDesktop:    DAMAGE extension not present
 XDesktop:    Will have to poll screen for changes
 XDesktop:    RANDR extension not present
 XDesktop:    Will not be able to handle session resize
 Main:        Listening on port 5900

and everything works, but not the way I would prefer it to.

**To Reproduce**
Attempt to run the above under CentOS 8

**Expected behavior**
With CentOS 7.5, running 'x0vncserver -display :0 -PasswordFile /etc/vncpasswd -AlwaysShared -SecurityTypes VncAuth,TLSVnc' works perfectly fine and with a VNC client one is presented with the typical user login screen. With CentOS 8, it doesn't.

**Screenshots**

**Client (please complete the following information):**
 - OS: CentOS 8
 - VNC client: Remmina
 - VNC client version: 1.4.12
 - Client downloaded from: snap list
 - remmina            v1.4.12+git4.0ad90e24       4745   latest/stable  remmina✓            -

**Server (please complete the following information):**
 - OS: CentOS 8
 - VNC server: Tigervnc
 - VNC server version: tigervnc-1.10.1-9.el8_3.x86_64
 - Server downloaded from: My Linux Distribution
 - Server was started using: 'x0vncserver -display :0 -PasswordFile /etc/vncpasswd -AlwaysShared -SecurityTypes VncAuth,TLSVnc'

**Additional context**
x0vncserver only works if AutomaticLogin=root. Specifying any other user results in 
No protocol specified
x0vncserver: unable to open display ":0"
although the non-root user is successfully logged in.

So, in summary, the only way I can get x0vncserver to work with CentOS 8 is to disable Wayland, and autologin root!

**Comment by CendioOssman** [comment]
Sounds like you've configured things correctly. But it also sounds like those settings haven't taken effect.

Could you include the complete contents of `custom.conf`?

And what does `sudo systemctl status gdm user-42.slice` look like? gdm needs to be visible on the monitor for this to be meaningful.

**Comment by mddflorida** [comment]
CendioOssman, thanks for the rapid response!

[root@mini-x ~]# **cat /etc/gdm/custom.conf** 
# GDM configuration storage
[daemon]
InitialSetupEnable=False
WaylandEnable=false
AutomaticLoginEnable=True
AutomaticLogin=root
[security]
[xdmcp]
[greeter]
[chooser]
[debug]
[root@mini-x ~]#  **systemctl status gdm user-42.slice**
● gdm.service - GNOME Display Manager
   Loaded: loaded (/usr/lib/systemd/system/gdm.service; enabled; vendor preset: enabled)
   Active: active (running) since Fri 2021-03-19 10:47:07 EDT; 11min ago
 Main PID: 5014 (gdm)
    Tasks: 3 (limit: 26213)
   CGroup: /system.slice/gdm.service
           └─5014 /usr/sbin/gdm

Mar 19 10:47:07 mini-x systemd[1]: Starting GNOME Display Manager...
Mar 19 10:47:07 mini-x systemd[1]: Started GNOME Display Manager.

● user-42.slice - User Slice of UID 42
   Loaded: loaded
  Drop-In: /usr/lib/systemd/system/user-.slice.d
           └─10-defaults.conf
   Active: inactive (dead)
[root@mini-x ~]#

**Comment by CendioOssman** [comment]
Hmm... that slice should not be empty. We should have a bunch of gdm stuff there. Are you sure gdm is shown on the display at this point? It automatically terminates stuff as soon as it isn't shown. So you'll unfortunately need to use a second machine and ssh in.

If that still doesn't show anything then please run `sudo systemctl status` and see if you can find gdm's stuff.

**Comment by mddflorida** [comment]
Cendio, yes I have two systems here with me. The mini-x system was logged in (AutomaticLogin=root). I removed AutomaticLoginEnable=True and AutomaticLogin=root from /etc/gdm/custom.conf and rebooted. The console has the normal GUI login screen, and the output of systemctl status gdm user-42.slice is loaded with output. [user42.txt](https://github.com/TigerVNC/tigervnc/files/6172493/user42.txt) but trying to run x0vncserver now gives me the original error messages: 
No protocol specified
x0vncserver: unable to open display ":0"

**Comment by mddflorida** [comment]
Cendio, is there anything else that I can do to provide more information for this? To me it appears to be a permissions problem, but I don't know where to look. Any assistance will be greatly appreciated.

**Comment by CendioOssman** [comment]
So from your dump we can see that an X server is indeed running. Unfortunately it is started with `-displayfd`, which makes it a bit tricky to determine which display it is running. What do you have in `/tmp/.X11-unix`? There should be a socket there for the server.

Also, which user are you running `x0vncserver` as?

**Comment by mddflorida** [comment]
I have an empty /tmp/.Test-unix directory...

[root@mini-x ~]# ls -la /tmp/.Test-unix/
total 4
drwxrwxrwt  2 root root    6 Mar 16 09:24 .
drwxrwxrwt 20 root root 4096 Mar 25 08:57 ..
[root@mini-x ~]#

x0vncserver is started by systemd
[root@mini-x ~]# cat /usr/lib/systemd/system/x0vncserver.service
[Unit]
Description=x0vncserver
After=display-manager.service
[Service]
Type=simple
RemainAfterExit=yes
ExecStart=/bin/bash -c 'x0vncserver -display :0 -PasswordFile /etc/vncpasswd -Log *:stderr:100 -AlwaysShared -SecurityTypes VncAuth,TLSVnc'
Restart=always
[Install]
WantedBy=graphical.target
[root@mini-x ~]#

[root@mini-x ~]# ps -ef | grep x0vncserver
root        2892       1  0 08:54 ?        00:00:00 x0vncserver -display :0 -PasswordFile /etc/vncpasswd -Log *:stderr:100 -AlwaysShared -SecurityTypes VncAuth,TLSVnc
[root@mini-x ~]#

**Comment by mddflorida** [comment]
My mistake from earlier today. Correction:

[root@mini-x ~]# ls -laR **/tmp/.X11-unix**
/tmp/.X11-unix:
total 4
drwxrwxrwt  2 root root   16 Mar 25 12:21 .
drwxrwxrwt 21 root root 4096 Mar 25 12:21 ..
srwxrwxrwx  1 root root    0 Mar 25 12:21 X0
[root@mini-x ~]# file /tmp/.X11-unix/X0 
/tmp/.X11-unix/X0: socket
[root@mini-x ~]#

**Comment by CendioOssman** [comment]
Hmm... So a socket seems to be in place. Exactly how are you starting `x0vncserver`?

And what do you get from a simple command such as `xdpyinfo`?

**Comment by mddflorida** [comment]
I start x0vncserver via the systemd service...
ExecStart=/bin/bash -c 'x0vncserver -display :0 -PasswordFile /etc/vncpasswd -Log *:stderr:100 -AlwaysShared -SecurityTypes VncAuth,TLSVnc'

I have also, as root, executed: 
[root@mini-x ~]# x0vncserver -display :0 -PasswordFile /etc/vncpasswd -Log *:stderr:100 -AlwaysShared -SecurityTypes VncAuth,TLSVnc 

Both methods exhibit the same behavior.

I ran the xdpyinfo command from a Gnome Terminal bash shell on the system console. The output 
is 1300+ lines so I attach the output for you. 
[root@mini-x ~]# wc -l /tmp/xdpwinfo.txt 
1333 /tmp/xdpwinfo.txt
[root@mini-x ~]#

[xdpwinfo.txt](https://github.com/TigerVNC/tigervnc/files/6213418/xdpwinfo.txt)

Thank you for your assistance.

**Comment by CendioOssman** [comment]
> I ran the xdpyinfo command from a Gnome Terminal bash shell on the system console. The output

I assume that means you logged in and ran this command? You need to start it the same way you start x0vncserver so we can compare.

**Comment by mddflorida** [comment]
Yes, I was auto logged in on the system console with the "Standard (X11 display server) on Xorg" selected, when I ran xdpyinfo above. I disabled auto login, rebooted, logged in as root and reran xdpyinfo. This time the only significant difference is:
**name of display:    :1**
version number:    11.0
vendor string:    The X.Org Foundation
vendor release number:    12008000
X.Org version: 1.20.8

Again, what I am trying to get working with CentOS 8 is to be able to VNC into the system and get the standard login screen. This worked with CentOS 7, but I am unable to get this working with CentOS 8.

**Comment by mddflorida** [comment]
I guess this is the end of the line on this problem.... So, here is what I have done to "work around" the problem. Firstly, I disabled Wayland in /etc/gdm/custom.conf and configured it to auto login root at boot time. Secondly, I set up systemd to start X0vncserver. The ONLY security is afforded by putting a password on the VNC connection. If you ever come up with something better, I'd really like to know. BTW, x11vnc suffers from the same problem with CentOS 8. Thanks for your time.

**Comment by mddflorida** [comment]
Cendio, thanks again for your time and efforts, but :0 functionality of x0vncserver that worked with CentOS 7.5 no longer works with CentOS 8. If you change your mind and reopen this ticket, I will be happy to assist with troubleshooting.

**Comment by CendioOssman** [comment]
Then we need to get that comparison in place. Exactly how were you starting `x0vncserver` on CentOS 7? What are the exact steps from a fresh boot of the server?

**Comment by mddflorida** [comment]
Cendio, with CentOS 7, I start x0vncserver at boot time via systemd.
# cat /usr/lib/systemd/system/x0vncserver.service
[Unit]
Description=x0vncserver
After=display-manager.service
[Service]
Type=simple
RemainAfterExit=yes
ExecStart=/bin/bash -c 'x0vncserver -display :0 -PasswordFile /etc/vncpasswd -AlwaysShared -SecurityTypes VncAuth,TLSVnc'
Restart=always
[Install]
WantedBy=graphical.target

**Comment by CendioOssman** [comment]
So that's what we'd like to try using `xdpyinfo` instead. Testing when logged in is not the same scenario so that doesn't really help us unfortunately.

**Comment by mddflorida** [comment]
Cendio, I am somewhat confused. How exactly should I proceed to run xdpyinfo to get the information you need? Should I disable x0vncserver.service and disable autologin of root, reboot then login at the console? Thank you!

**Comment by mddflorida** [comment]
I ran xdpyinfo on both a 7.5 and an 8.0 CentOS system. The hardware/motherbaord of each is identical. Neither system was configured for autologin. I logged in as root at the system console for each.

For 7.5: "name of display:    :0"
For 8.0: "name of display:    :1.0"

Attached are xdpyinfo output for both systems.

[CentOS_8.0_xdpyinfo.txt](https://github.com/TigerVNC/tigervnc/files/6266896/CentOS_8.0_xdpyinfo.txt)
[CentOS_7.5_xdpyinfo.txt](https://github.com/TigerVNC/tigervnc/files/6266897/CentOS_7.5_xdpyinfo.txt)

**Comment by CendioOssman** [comment]
> Cendio, I am somewhat confused. How exactly should I proceed to run xdpyinfo to get the information you need? Should I disable x0vncserver.service and disable autologin of root, reboot then login at the console? Thank you!

Pretty much. Create a service just like you did for `x0vncserver` so that it starts the same way. Then check the logs for the service after disabling autologin and rebooting.

**Comment by mddflorida** [comment]
I disabled autologin and restarted the system. On reboot, the standard graphical login screen appears on the console as expected. I hope the attached files help narrow down the problem. Thanks!

[etc_gdm_custom.conf.txt](https://github.com/TigerVNC/tigervnc/files/6287679/etc_gdm_custom.conf.txt)
[systemctl_status_x0vncserver.service.txt](https://github.com/TigerVNC/tigervnc/files/6287680/systemctl_status_x0vncserver.service.txt)
[var_log_messages.txt](https://github.com/TigerVNC/tigervnc/files/6287681/var_log_messages.txt)
[x0vncserver.service.txt](https://github.com/TigerVNC/tigervnc/files/6287692/x0vncserver.service.txt)

**Comment by CendioOssman** [comment]
I'm afraid not. You need to create a second service and run `xdpyinfo` in that instead of `x0vncserver` so we can see if it has the same issues.

**Comment by mddflorida** [comment]
Sorry for my delay with replying, but I have been pulled in many directions.... I ran xdpyinfo &> /tmp/xdpyinfo.txt, using rc.local, with and without autologin. In both cases the output was "xdpyinfo:  unable to open display ""." Maybe I'm doing something wrong??? Please advise. Thanks!

**Comment by CendioOssman** [comment]
Great. So that tells us there isn't something wrong with specifically `x0vncserver`, but rather with something general. I'm guessing they've locked down GDM more in modern versions, so there are some extra steps needed.

Unfortunately I'm no expert on GDM so I'm not sure exactly what needs to be changed. I suspect the issue has to do with XAUTH. Have you seen any settings in GDM regarding that? Or regarding xhost?

You could also try running `x0vncserver` as the same user as gdm. Right now I assume you run it as root?

**Comment by mddflorida** [comment]
I am no expert on GDM either. I have spent many hours seeking a solution to this, and I'm not the only one either. Smarter people will need to dig into this I guess. In the meantime, I have a hack that I can use although I don't like it. The question remains, "How do I get VNC working on the login screen?" Yes, I run X0vncserver as root and it is started via systemd x0vncserver.service.

Any ideas you may have will be greatly appreciated. Once again, thank you for your time. I'll be standing by to help troubleshoot this.

**Comment by mikaku** [comment]
> You could also try running `x0vncserver` as the same user as gdm. Right now I assume you run it as root?

That did the trick! :smiley: 
I had a similar problem with `x11vnc` on a CentOS 8.4 and executing it with the same user as `gdm` worked like a charm:

```
sudo -u gdm x11vnc -display :0 -auth /run/user/42/gdm/Xauthority
```
